File: neet_code_all_cpp/bin_search/find_first_and_last_in_sorted_array.cpp
#include <cwchar>
#include <iostream>
#include <vector>

class Solution {
    public:
        // AC
        std::vector<int> searchRange(std::vector<int>& nums, int target) {
            auto [left, right] = std::pair(0, (int)nums.size() - 1);
            while(left <= right) {
                int mid = left + (right - left) / 2;

                if(nums[mid] == target) {
                    auto [l_mid, r_mid] = std::pair(mid, mid);

                    while(l_mid - 1 >= 0 && nums[l_mid - 1] == target) {
                        --l_mid;
                    }
                    while(r_mid + 1 < (int)nums.size() && nums[r_mid + 1] == target) {
                        ++r_mid;
                    }

                    return std::vector{l_mid, r_mid};

                }

                if(nums[mid] < target) {
                    left = mid + 1;
                } else if(nums[mid] > target) {
                    right = mid - 1;
                }
            }

            return std::vector{-1, -1};
        }
};

// 模範解答
class SolutionAns {
    public:
        std::vector<int> searchRange(std::vector<int>& nums, int target) {
            if(nums.empty()) {
                return {-1, -1};
            }
            auto [left, right] = std::pair(0, (int)nums.size() - 1);
            while(left < right) {
                int mid = left + (right - left) / 2;
                if(nums[mid] >= target) {
                    right = mid;
                } else {
                    left = mid + 1;
                }
            }
            if(nums[left] != target) {
                return {-1, -1};
            }
            int left_result = left;
            left = 0;
            right = nums.size() - 1;

            while(left < right) {
                int mid = left + (right - left + 1) / 2;
                if(nums[mid] <= target) {
                    left = mid;
                } else {
                    right = mid - 1;
                }
            }

            return {left_result, right};
        }
};

void show_res(std::vector<int>&& res) {
    std::cout << res[0] << " " << res[1] << std::endl;
}

int main(void) {
    using Case = std::pair<std::vector<int>, int>;
    Case case_1 = {{5, 7, 7, 8, 8, 10}, 8};
    // => [3, 4]
    Case case_2 = {{5, 7, 7, 8, 8, 10}, 6};
    // => [-1, -1]
    Case case_3 = {{}, 0};
    // => [-1, -1]

    Solution s_1;
    show_res(s_1.searchRange(case_1.first, case_1.second));
    show_res(s_1.searchRange(case_2.first, case_2.second));
    show_res(s_1.searchRange(case_3.first, case_3.second));

    SolutionAns s_ans;
    show_res(s_ans.searchRange(case_1.first, case_1.second));
    show_res(s_ans.searchRange(case_2.first, case_2.second));
    show_res(s_ans.searchRange(case_3.first, case_3.second));
}

File: neet_code_all/src/bin_search/find_first_and_last_in_sorted_array.rs
struct Solution {}
impl Solution {
    // AC
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let (mut left, mut right) = (0, nums.len() as i32 - 1);
        while left <= right {
            let mid = left + (right - left) / 2;

            if nums[mid as usize] == target {
                //println!("mid: {mid}");
                let (mut l_mid, mut r_mid) = (mid, mid);
                while l_mid - 1 >= 0 && nums[l_mid as usize - 1] == target {
                    l_mid -= 1;
                }
                while r_mid + 1 < nums.len() as i32 && nums[r_mid as usize + 1] == target {
                    r_mid += 1;
                }

                return vec![l_mid, r_mid];
            }

            if nums[mid as usize] < target {
                left = mid + 1;
            } else if nums[mid as usize] > target {
                right = mid - 1;
            }
        }

        vec![-1, -1]
    }
}

// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    // AC
    pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
        if nums.is_empty() {
            return vec![-1, -1];
        }
        let (mut left, mut right) = (0, nums.len() as i32 - 1);
        while left < right {
            let mid = left + (right - left) / 2;
            if nums[mid as usize] >= target {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        if nums[left as usize] != target {
            return vec![-1, -1];
        }

        let left_result = left;
        left = 0;
        right = nums.len() as i32 - 1;

        while left < right {
            let mid = left + (right - left + 1) / 2;
            if nums[mid as usize] <= target {
                left = mid;
            } else {
                right = mid - 1;
            }
        }

        vec![left_result, right]
    }
}

fn main() {
    let case_1 = (vec![5, 7, 7, 8, 8, 10], 8);
    // => [3, 4]
    let case_2 = (vec![5, 7, 7, 8, 8, 10], 6);
    // => [-1, -1]
    let case_3 = (Vec::<i32>::new(), 0);
    // => [-1, -1]

    println!(
        "case_1: {:?}",
        Solution::search_range(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        Solution::search_range(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        Solution::search_range(case_3.0.clone(), case_3.1)
    );

    println!(
        "case_1: {:?}",
        SolutionAnsCpp::search_range(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {:?}",
        SolutionAnsCpp::search_range(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {:?}",
        SolutionAnsCpp::search_range(case_3.0.clone(), case_3.1)
    );
}
